Neos.EventSourcedNeosAdjustments/Classes/EventSourcedRouting/Projection/DocumentUriPathFinder.php

```diff
@@ -0,0 +1,146 @@
+    }
+
+    /**
+     * @param NodeAggregateIdentifier $parentNodeAggregateIdentifier
+     * @param string $dimensionSpacePointHash
+     * @return DocumentNodeInfo
+     * @throws NodeNotFoundException
+     */
+    public function getLastChildNode(NodeAggregateIdentifier $parentNodeAggregateIdentifier, string $dimensionSpacePointHash): DocumentNodeInfo
+    {
+        return $this->fetchSingle('dimensionSpacePointHash = :dimensionSpacePointHash AND parentNodeAggregateIdentifier = :parentNodeAggregateIdentifier AND succeedingNodeAggregateIdentifier IS NULL', compact('dimensionSpacePointHash', 'parentNodeAggregateIdentifier'));
+    }
+
+    /**
+     * @param NodeAggregateIdentifier $nodeAggregateIdentifier
+     * @return \Iterator|DocumentNodeInfo[]
+     */
+    public function getNodeVariantsById(NodeAggregateIdentifier $nodeAggregateIdentifier): \Iterator
+    {
+        try {
+            $iterator = $this->dbal->executeQuery('SELECT * FROM ' . DocumentUriPathProjector::TABLE_NAME_DOCUMENT_URIS . ' WHERE nodeAggregateIdentifier = :nodeAggregateIdentifier', ['nodeAggregateIdentifier' => $nodeAggregateIdentifier]);
+        } catch (DBALException $e) {
+            throw new \RuntimeException(sprintf('Failed to get node variants for id "%s": %s', $nodeAggregateIdentifier, $e->getMessage()), 1599665543, $e);
+        }
+        foreach ($iterator as $nodeSource) {
+            yield DocumentNodeInfo::fromDatabaseRow($nodeSource);
+        }
+    }
+
+    public function getLiveContentStreamIdentifier(): ContentStreamIdentifier
+    {
+        if ($this->liveContentStreamIdentifierRuntimeCache === null) {
+            try {
+                $contentStreamIdentifier = $this->dbal->fetchColumn('SELECT contentStreamIdentifier FROM ' . DocumentUriPathProjector::TABLE_NAME_LIVE_CONTENT_STREAMS . ' LIMIT 1');
+            } catch (DBALException $e) {
+                throw new \RuntimeException(sprintf('Failed to fetch contentStreamIdentifier for live workspace: %s', $e->getMessage()), 1599666764, $e);
+            }
+            if (!is_string($contentStreamIdentifier)) {
+                throw new \RuntimeException('Failed to fetch contentStreamIdentifier for live workspace, probably you have to replay the "documenturipath" projection', 1599667894);
+            }
+            $this->liveContentStreamIdentifierRuntimeCache = ContentStreamIdentifier::fromString($contentStreamIdentifier);
+        }
+        return $this->liveContentStreamIdentifierRuntimeCache;
+    }
+
+    /**
+     * @param string $where
+     * @param array $parameters
+     * @return DocumentNodeInfo
+     * @throws NodeNotFoundException
+     */
+    private function fetchSingle(string $where, array $parameters): DocumentNodeInfo
+    {
+        try {
+            $row = $this->dbal->fetchAssoc('SELECT * FROM ' . DocumentUriPathProjector::TABLE_NAME_DOCUMENT_URIS . ' WHERE ' . $where . ' LIMIT 1', $parameters, DocumentUriPathProjector::COLUMN_TYPES_DOCUMENT_URIS);
```

Comment:
is this "limit 1" safe here; or should we throw exceptions if we read more than one row?

Reply:
`Connection::fetchAssoc()` only returns the first result anyways, the `LIMIT 1` is just a performance optimization to tell the db server to stop looking.
Probably it's not relevant because all calls to this method include unique constraints, but it shouldn't do any harm.
I added a comment